{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Importing libraries\n",
    "import os\n",
    "import sys\n",
    "import json\n",
    "import numpy as np\n",
    "import modin.pandas as pd\n",
    "import requests as rq\n",
    "from snps import SNPs\n",
    "import flatten_json as fj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Constants of the application\n",
    "os.environ[\"MODIN_ENGINE\"] = \"dask\" \n",
    "REFSNP_API = 'https://api.ncbi.nlm.nih.gov/variation/v0/refsnp/{snp_id}/frequency'\n",
    "OPENSNP_API = 'https://opensnp.org/data/{user_id}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Module functions\n",
    "\n",
    "## Safe URL API requests from https://github.com/ncbi/dbsnp/blob/master/tutorials/Variation%20Services/Jupyter_Notebook/spdi_batch.ipynb\n",
    "def api_request(url):\n",
    "    try:\n",
    "        r = rq.get(url)\n",
    "    except rq.exceptions.Timeout:\n",
    "        # Maybe set up for a retry, or continue in a retry loop\n",
    "        print(\"ERROR: Timeout\")\n",
    "    except rq.exceptions.TooManyRedirects:\n",
    "        # Tell the user their URL was bad and try a different one\n",
    "        print(\"ERROR: bad url =\" + url)\n",
    "    except rq.exceptions.RequestException as e:\n",
    "        # catastrophic error. bail.\n",
    "        print(e)\n",
    "        sys.exit(1)\n",
    "    if (r.status_code == 200):\n",
    "        return r\n",
    "    else:\n",
    "        print(\"ERROR: status code = \" + str(r.status_code))\n",
    "        return None\n",
    "\n",
    "## Recovering the reference allele of the genotyped variant\n",
    "def get_reference_allele_api(snp_id):\n",
    "    var_dict = api_request(REFSNP_API.format(snp_id=snp_id)).json()\n",
    "    allele_freqs = {}\n",
    "\n",
    "    var_dict_flatten = fj.flatten(var_dict)\n",
    "    for key, value in var_dict_flatten.items():\n",
    "        if('allele_counts' in key):\n",
    "            allele = key[-1]\n",
    "            if(allele in allele_freqs.keys()):\n",
    "                allele_freqs[allele] = allele_freqs[allele] + value\n",
    "            else:\n",
    "                allele_freqs[allele] = value\n",
    "\n",
    "    major_allele = sorted(allele_freqs.items(), key=lambda item: item[1])[-1][0]\n",
    "    print(\"SNP: {snp_id} has a major allele of {major_allele}\".format(snp_id=snp_id, major_allele=major_allele))\n",
    "    return major_allele\n",
    "\n",
    "## Recovering the genotype dataset from a given individual\n",
    "def get_genotype_individual(user_id):\n",
    "    if(user_id != None):\n",
    "        raw_data = api_request(OPENSNP_API.format(user_id=user_id))\n",
    "        indiv_snps = SNPs(raw_data.content)\n",
    "        genotype = pd.DataFrame(indiv_snps.snps.reset_index())\n",
    "    return genotype\n",
    "\n",
    "## Extracting the genotyped alleles from each individual for the PRS\n",
    "def get_genotyped_alleles_individual(user_genotype, model_table):\n",
    "    merged_df = user_genotype.merge(model_table,left_on='rsid', right_on='snp_id', how='inner')\n",
    "    return merged_df\n",
    "\n",
    "## Takes a model table and extracts the \n",
    "def assign_reference_allele(snp_entry):\n",
    "    snp_id = snp_entry['snp_id'].replace(r'rs', '')\n",
    "    ref_allele = get_reference_allele_api(snp_id)\n",
    "    return ref_allele\n",
    "\n",
    "## Get minor allele dosage\n",
    "def get_minor_allele_dosage(snp_entry):\n",
    "    reference_allele = snp_entry['reference_allele']\n",
    "    genotype = snp_entry['genotype']\n",
    "    dosage = 2 - genotype.count(reference_allele)\n",
    "    return dosage\n",
    "\n",
    "## Obtaining the reference (no-effect) allele\n",
    "def get_reference_alleles_model(model_table):\n",
    "    model_table['reference_allele'] = model_table.apply(lambda x: assign_reference_allele(x), axis=1)\n",
    "    return model_table\n",
    "\n",
    "## Attaching the reference (no-effect) allele and the dosage of the minor (effect) allele\n",
    "def merge_allelic_data(user_genotype, model_table):\n",
    "    final_genotype_individual = get_genotyped_alleles_individual(user_genotype, model_table)\n",
    "    final_genotype_individual['minor_allele_dosage'] = final_genotype_individual.apply(lambda x: get_minor_allele_dosage(x), axis=1)\n",
    "    return final_genotype_individual\n",
    "\n",
    "## Calculating the PRS from the effect size and the minor (effect) allele dosage\n",
    "def calculating_prs_individual(final_genotype_individual):\n",
    "    prs = final_genotype_individual['effect_size'] * final_genotype_individual['minor_allele_dosage']\n",
    "    return sum(prs)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "UserWarning: Dask execution environment not yet initialized. Initializing...\n",
      "To remove this warning, run the following python code before doing dataframe operations:\n",
      "\n",
      "    from distributed import Client\n",
      "\n",
      "    client = Client()\n",
      "\n",
      "UserWarning: Distributing <class 'pandas.core.frame.DataFrame'> object. This may take some time.\n",
      "UserWarning: User-defined function verification is still under development in Modin. The function provided is not verified.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The PRS for the Model A is: -0.6663467602183987\n",
      "The PRS for the Model B is: -0.7662909745323819\n"
     ]
    }
   ],
   "source": [
    "## Getting the individual genotypes\n",
    "individuo = get_genotype_individual('11008.23andme.9131')\n",
    "\n",
    "mod1 = pd.read_csv('datasets/model_a.tsv', sep='\\t')\n",
    "mod1 = get_reference_alleles_model(mod1)\n",
    "geno_indiv_mod1 = merge_allelic_data(individuo, mod1)\n",
    "prs_mod1 = calculating_prs_individual(geno_indiv_mod1)\n",
    "print(\"The PRS of the individual {} for the Model A is: {prs}\".format(prs=prs_mod1))\n",
    "\n",
    "mod2 = pd.read_csv('datasets/model_b.tsv', sep='\\t')\n",
    "mod2 = get_reference_alleles_model(mod2)\n",
    "geno_indiv_mod2 = merge_allelic_data(individuo, mod2)\n",
    "prs_mod2 = calculating_prs_individual(geno_indiv_mod2)\n",
    "print(\"The PRS for the Model B is: {prs}\".format(prs=prs_mod2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>genotype</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>minor_allele_dosage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs2001142</td>\n",
       "      <td>1</td>\n",
       "      <td>7348943</td>\n",
       "      <td>CC</td>\n",
       "      <td>rs2001142</td>\n",
       "      <td>0.008742</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1009941</td>\n",
       "      <td>1</td>\n",
       "      <td>9275194</td>\n",
       "      <td>AG</td>\n",
       "      <td>rs1009941</td>\n",
       "      <td>-0.002960</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs2990678</td>\n",
       "      <td>1</td>\n",
       "      <td>13837213</td>\n",
       "      <td>CT</td>\n",
       "      <td>rs2990678</td>\n",
       "      <td>-0.006790</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs12751422</td>\n",
       "      <td>1</td>\n",
       "      <td>27493679</td>\n",
       "      <td>TT</td>\n",
       "      <td>rs12751422</td>\n",
       "      <td>-0.006500</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs11583644</td>\n",
       "      <td>1</td>\n",
       "      <td>30419370</td>\n",
       "      <td>AG</td>\n",
       "      <td>rs11583644</td>\n",
       "      <td>0.013899</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>rs4811971</td>\n",
       "      <td>20</td>\n",
       "      <td>56796784</td>\n",
       "      <td>GG</td>\n",
       "      <td>rs4811971</td>\n",
       "      <td>0.026296</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>rs1918388</td>\n",
       "      <td>21</td>\n",
       "      <td>23964448</td>\n",
       "      <td>AA</td>\n",
       "      <td>rs1918388</td>\n",
       "      <td>-0.010879</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>rs2827703</td>\n",
       "      <td>21</td>\n",
       "      <td>24143818</td>\n",
       "      <td>GG</td>\n",
       "      <td>rs2827703</td>\n",
       "      <td>-0.006051</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>rs4819996</td>\n",
       "      <td>22</td>\n",
       "      <td>17790731</td>\n",
       "      <td>AG</td>\n",
       "      <td>rs4819996</td>\n",
       "      <td>0.001344</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>rs140161</td>\n",
       "      <td>22</td>\n",
       "      <td>23845547</td>\n",
       "      <td>AG</td>\n",
       "      <td>rs140161</td>\n",
       "      <td>-0.026361</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>201 rows x 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsid chrom       pos genotype      snp_id  effect_size  \\\n",
       "0     rs2001142     1   7348943       CC   rs2001142     0.008742   \n",
       "1     rs1009941     1   9275194       AG   rs1009941    -0.002960   \n",
       "2     rs2990678     1  13837213       CT   rs2990678    -0.006790   \n",
       "3    rs12751422     1  27493679       TT  rs12751422    -0.006500   \n",
       "4    rs11583644     1  30419370       AG  rs11583644     0.013899   \n",
       "..          ...   ...       ...      ...         ...          ...   \n",
       "196   rs4811971    20  56796784       GG   rs4811971     0.026296   \n",
       "197   rs1918388    21  23964448       AA   rs1918388    -0.010879   \n",
       "198   rs2827703    21  24143818       GG   rs2827703    -0.006051   \n",
       "199   rs4819996    22  17790731       AG   rs4819996     0.001344   \n",
       "200    rs140161    22  23845547       AG    rs140161    -0.026361   \n",
       "\n",
       "    reference_allele  minor_allele_dosage  \n",
       "0                  C                    0  \n",
       "1                  A                    1  \n",
       "2                  C                    1  \n",
       "3                  C                    2  \n",
       "4                  A                    1  \n",
       "..               ...                  ...  \n",
       "196                A                    2  \n",
       "197                C                    2  \n",
       "198                G                    0  \n",
       "199                G                    1  \n",
       "200                G                    1  \n",
       "\n",
       "[201 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geno_indiv_mod1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.7 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "6b2ed07406f43fefe29d7e6a5344fe55f2939fdcf75c25060c3ecc711fe31b42"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
